src/main/java/io/projectreactor/ModuleUtils.java

```diff
@@ -0,0 +1,125 @@
+package io.projectreactor;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.constructor.Constructor;
+
+import org.springframework.core.io.ClassPathResource;
+
+import reactor.core.Exceptions;
+import reactor.core.publisher.Mono;
+import reactor.netty.http.client.HttpClient;
+import reactor.util.Logger;
+import reactor.util.Loggers;
+
+/**
+ * @author Simon Baslé
+ */
+public class ModuleUtils {
+
+	private static final Logger LOGGER = Loggers.getLogger(ModuleUtils.class);
+
+	public static void loadModulesFromYmlInto(ClassPathResource resource, Map<String, Module> target) {
+		Yaml yaml = new Yaml(new Constructor(Module.class));
+		try {
+			yaml.loadAll(resource.getInputStream()).forEach(o -> {
+				Module module = (Module)o;
+				module.sortAndDeduplicateVersions();
+				target.put(module.getName(), module);
+			});
+		}
+		catch (IOException e) {
+			throw new UncheckedIOException(e);
+		}
+	}
+
+	public static void fetchVersionsFromArtifactory(Map<String, Module> modules, String... moduleNames) {
+		final HttpClient client = HttpClient.create().baseUrl("https://repo.spring.io/api/search");
+		final String repos = "&repos=snapshot,milestone,release";
+
+		for (String moduleName : moduleNames) {
+			Module module = modules.get(moduleName);
+			if (module == null) continue;
+
+			final String params = "/versions?g=" + module.getGroupId() + "&a=" + module.getArtifactId() + repos;
+			LOGGER.info("Loading version information for {} via GET {}", moduleName, params);
+
+			client.get()
+			      .uri(params)
+			      .response((r, content) -> {
+				      if (r.status().code() < 400) {
+					      return content.aggregate()
+					                    .asString()
+					                    .doOnNext(json -> loadModuleVersionsFromArtifactoryVersionsSearch(json, module));
+				      }
+				      else {
+					      return content.aggregate()
+					                    .asString()
+					                    .flatMap(body -> Mono.error(new RuntimeException(r.status() + ": " + body))
+					                    );
+				      }
+			      })
+			      .doOnError(e -> LOGGER.warn("Couldn't scrape versions for {}: {}", moduleName, e.getMessage()))
+			      .onErrorReturn("")
+			      .blockLast();
```

Comment:
could use a flatMap ;)

Reply:
not sure that's what you intended but I simplified the logic a bit by avoiding the `Mono.error` -> `doOnError().onErrorReturn("")` dance.

Reply:
my idea was to not block in every iteration but instead use a Flux of moduleNames and `flatMap` the fetching (so that we concurrently run it instead of one-by-one, blocking style)

Reply:
ah got it, updated :+1: